x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java

```diff
@@ -427,4 +434,43 @@ public static boolean isPushableFieldAttribute(Expression exp, Predicate<FieldAt
         return false;
     }
 
+    private static class SpatialFromDocValuesExtraction extends OptimizerRule<AggregateExec> {
+        @Override
+        protected PhysicalPlan rule(AggregateExec aggregate) {
+            var foundAttribute = new Holder<Attribute>(null);
+
+            PhysicalPlan plan = aggregate.transformDown(UnaryExec.class, exec -> {
+                if (exec instanceof AggregateExec agg) {
+                    var orderedAggregates = new ArrayList<NamedExpression>();
+                    var changedAggregates = false;
+                    for (NamedExpression aggExpr : agg.aggregates()) {
+                        if (aggExpr instanceof Alias as && as.child() instanceof SpatialAggregateFunction af) {
+                            // Firstly we tell the aggregation function that is should expect doc-values instead of source values
+                            var changed = as.replaceChild(af.withDocValues());
+                            changedAggregates = true;
+                            if (af.field() instanceof Attribute fieldAttribute) {
+                                foundAttribute.set(fieldAttribute);
```

Comment:
Just a double-check: are you taking this recent change into consideration? https://github.com/elastic/elasticsearch/pull/104387
(I don't have a real use case where this could make a big difference, just wondering if you checked it)

Reply:
Looking at that now. It turns out that if we nest a function inside the `st_centroid` then we won't load from doc-values. The queries work, but the optimization is not activated. I'm busy debugging that now to see if there is a simple solution, but this could also be a followup PR to enabled this case.

Reply:
I did need to make a small change to the local physical planner to get all tests to pass with nested functions. They are not using the doc-values optimization, but at least the queries run with nested fields.